test/test_mps.py

```diff
@@ -1383,6 +1383,88 @@ def test_randperm(self, device="mps"):
             torch.randperm(n, out=non_contiguous_tensor)
             self.assertEqual(res.cpu().sort().values.long(), torch.arange(n, device=device))
 
+    # For some reason, max_pool3d's backward pass for MPS has a precision
+    # problem. This test is a smaller version of the failing test:
+    #   ```
+    #   PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=60 \
+    #   python test/test_mps.py TestConsistencyMPS.test_output_grad_match_nn_functional_max_pool3d_mps_float16
+    #   ````
+    # TODO: Remove once this issue is fixed
+    def test_max_pool3d_backward_precision_float16(self):
+        input_size = (1, 1, 1, 5, 5)
+        kwargs = {
+            'kernel_size': (1, 3, 3),
+            'stride': 1,
+            'ceil_mode': False,
+            'padding': (0, 1, 1),
+            'dilation': 1,
+            'return_indices': True,
+        }
+        dtype = torch.float16
+
+        # These are the tolerances used in test_output_grad_match
+        atol = 1e-5
+        rtol = 1e-3
+
+        input_cpu = torch.tensor(
+            [[[
+                [0.6143, 0.6294, 0.4937, 0.1738, 0.0171],
+                [0.3511, 0.2036, 0.3657, 0.3608, 0.6421],
+                [0.3384, 0.7241, 0.4609, 0.2085, 0.2344],
+                [0.3384, 0.6514, 0.4692, 0.1621, 0.5977],
+                [0.6113, 0.2598, 0.4917, 0.0693, 0.3472]
+            ]]],
+            dtype=dtype)
+
+        input_mps = input_cpu.to('mps').detach()
+        input_cpu.requires_grad_(True)
+        input_mps.requires_grad_(True)
+
+        # Forward call results match
+        output_cpu, indices_cpu = torch.nn.functional.max_pool3d(input_cpu, **kwargs)
+        output_mps, indices_mps = torch.nn.functional.max_pool3d(input_mps, **kwargs)
+        self.assertEqual(output_cpu, output_mps, atol=atol, rtol=rtol)
+        self.assertEqual(indices_cpu, indices_mps, atol=atol, rtol=rtol)
+
+        output_grad_cpu = torch.tensor(
+            [[[
+                [0.0718, 0.6611, 0.3027, 0.6875, 0.5610],
+                [0.1816, 0.5708, 0.6470, 0.9761, 0.9067],
+                [0.7988, 0.1284, 0.0864, 0.6509, 0.2246],
+                [0.1357, 0.2300, 0.3999, 0.8418, 0.1621],
+                [0.9526, 0.2236, 0.8101, 0.5869, 0.9888]
+            ]]],
+            dtype=dtype)
+        output_grad_mps = output_grad_cpu.to('mps').detach()
+
+        # Backward call
+        output_cpu.backward(output_grad_cpu)
+        output_mps.backward(output_grad_mps)
+
+        # The input grads for CPU and MPS should be equal, but in this case they
+        # are not.
+        self.assertNotEqual(input_mps.grad, input_cpu.grad, atol=atol, rtol=rtol)
+
+        # This is the mismatching element
+        self.assertEqual(input_cpu.grad[0, 0, 2, 1], 3.1816, atol=atol, rtol=rtol)
+        self.assertEqual(input_mps.grad[0, 0, 2, 1], 3.1758, atol=atol, rtol=rtol)
```

Comment:
To me, this seems like an abnormally large difference. The elements getting summed from the output grad here are: `[0.1816, 0.5708, 0.6470, 0.7988, 0.1284, 0.0864, 0.1357, 0.2300, 0.3999]`. Summing just nine numbers of similar magnitudes should not give such a high error, even considering if the order summation differs

Reply:
Interestingly, `torch.sum()` gives the same result whether it's run on MPS and CPU tensors, even if the order changes. But it's yet a different result than `max_pool3d_backward` for both MPS and CPU.

```
>>> a = torch.tensor([0.1816, 0.5708, 0.6470, 0.7988, 0.1284, 0.0864, 0.1357, 0.2300, 0.3999], dtype=torch.float16)
>>> a[torch.randperm(9)].contiguous().sum()
tensor(3.1797, dtype=torch.float16)
>>> a[torch.randperm(9)].contiguous().sum()
tensor(3.1797, dtype=torch.float16)
>>> a.to('mps')[torch.randperm(9)].contiguous().sum()
tensor(3.1797, device='mps:0', dtype=torch.float16)
>>> a.to('mps')[torch.randperm(9)].contiguous().sum()
tensor(3.1797, device='mps:0', dtype=torch.float16)
```

Reply:
I've updated the tolerance values. If this is alright, I'll remove the test I added here as well before merging

Reply:
I decided to get rid of this test now because it was making CI fail, even though it passes when I run it locally, and I only added this test with the intention of removing it after understanding why CPU and MPS can give such different results